### PR TITLE
[uss_qualifier] netrid: DSS0020 - check DSS endpoints are encrypted

### PR DESCRIPTION
**Pull request by Shastick**
This new scenario checks that port 80 of a DSS's base URL either:
 - refuses connections
 - answers with a redirect to the https base url

The scenario also checks that the https port is available to receive requests. Currently there is no check to enforce a minimal cipher strength.

Note: this scenario requires a DSS configured with an `https://` base url, meaning that locally and in CI, it will be skipped.

Part of the effort on #754

### CODE REVIEW

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        parsed_url = urlparse(self._dss.base_url)
+        hostname = parsed_url.hostname
+
+        if hostname is not None:
+            with self.check(
+                "Connection to HTTP port fails or redirects to HTTPS port",
+                self._dss.participant_id,
+            ) as check:
+                try:
+                    response = requests.get(
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py:54]: For background, the requirement states:

> Communication between a USS and DSS instances shall (DSS0020) be encrypted using an industry-standard encryption mechanism with a minimum encryption strength of 128 bits.

Here I interpret that _anything else_ than a redirection is a form of communication, and that if it happens over HTTP, then the requirement is failed. Hence a query to _any_ endpoint is fair game.

I also agree that issuing a query that we would expect would succeed over https would make sense, and we can probably simply do both.

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                        f"http://{hostname}/{parsed_url.path}",
+                        timeout=10,
+                        allow_redirects=False,
+                    )
+                    # If we can connect, we want to check that we are being redirected:
+                    if response.status_code not in [301, 308]:
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py:60]: 302 and 307 are typically for temporary redirects, so I felt like they should not count, but there's no real strong reason for that. I'll extend this to include 302 and 307

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        parsed_url = urlparse(self._dss.base_url)
+        hostname = parsed_url.hostname
+
+        if hostname is not None:
+            with self.check(
+                "A request can be sent over HTTPS",
+                self._dss.participant_id,
+            ) as check:
+                try:
+                    requests.get(
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py:102]: The way I see it, is that the OpenAPI spec defines an _application protocol_, and that here we are testing something about the _transport layer_.

At the level we are testing we don't really care about what happens at the application level: we're expecting the socket to open successfully and that's it.

File: monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py:54]: It should also be noted that DSS0020 has nothing to do with the OpenAPI Spec (it does not mention it).

[the remark this answers is not in the document]
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py]: > Hence a query to any endpoint is fair game.

My point was that the raw base URL is actually *not* an endpoint. It's a prefix to the endpoints.
Depending on how the infra around the DSS is deployed, the request might not even reach at all the DSS, e.g. if you have an ingress/RP that will forward calls only for specific paths.

[the remark this answers is not in the document]
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common/dss/endpoint_encryption.py]: > Hence a query to any endpoint is fair game.

My point was that the raw base URL is actually not an endpoint. It's a prefix to the endpoints.
Depending on how the infra around the DSS is deployed, the request might not even reach at all the DSS, e.g. if you have an ingress/RP that will forward calls only for specific paths.

**Review by mickmis — APPROVED** [review]
I've reworked the scenario to simplify it after the discussions regarding hitting an endpoint or not.
Since it cannot run in the CI, I tested it locally by hard-coding URLs `http://github.com` and `https//github.com` in the python scenario, and it behaves as expected.

**Pull request merged**